reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/job/ReportsSchedulerJobRunner.java

```diff
@@ -0,0 +1,221 @@
+import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A sample job runner class.
+ *
+ * The job runner should be a singleton class if it uses Elasticsearch client or other objects passed
+ * from Elasticsearch. Because when registering the job runner to JobScheduler plugin, Elasticsearch has
+ * not invoke plugins' createComponents() method. That is saying the plugin is not completely initalized,
+ * and the Elasticsearch {@link org.elasticsearch.client.Client}, {@link ClusterService} and other objects
+ * are not available to plugin and this job runner.
+ *
+ * So we have to move this job runner intialization to {@link Plugin} createComponents() method, and using
+ * singleton job runner to ensure we register a usable job runner instance to JobScheduler plugin.
+ *
+ * This sample job runner takes the "indexToWatch" from job parameter and logs that index's shards.
+ */
+public class ReportsSchedulerJobRunner implements ScheduledJobRunner {
+
+    private static final Logger log = LogManager.getLogger(ScheduledJobRunner.class);
+    private static final String JOB_QUEUE_INDEX_NAME = ".reports_scheduler_job_queue";
+
+    private static ReportsSchedulerJobRunner INSTANCE;
+
+    public static ReportsSchedulerJobRunner getJobRunnerInstance() {
+        if(INSTANCE != null) {
+            return INSTANCE;
+        }
+        synchronized (ReportsSchedulerJobRunner.class) {
+            if(INSTANCE != null) {
+                return INSTANCE;
+            }
+            INSTANCE = new ReportsSchedulerJobRunner();
+            return INSTANCE;
+        }
+    }
+
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+    private Client client;
+
+    private ReportsSchedulerJobRunner() {
+        // Singleton class, use getJobRunner method instead of constructor
+    }
+
+    public void setClient(Client client) {
+        this.client = client;
+    }
+
+    public void setClusterService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+    public void setThreadPool(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+    }
+
+
+    @Override
+    public void runJob(ScheduledJobParameter jobParameter, JobExecutionContext context) {
+        if(!(jobParameter instanceof JobParameter)) {
+            throw new IllegalStateException("Job parameter is not instance of JobParameter, type: "
+                    + jobParameter.getClass().getCanonicalName());
+        }
+
+        if(this.clusterService == null) {
+            throw new IllegalStateException("ClusterService is not initialized.");
+        }
+
+        if (this.threadPool == null) {
+            throw new IllegalStateException("ThreadPool is not initialized.");
+        }
+
+        Runnable runnable = () -> {
+
+            JobParameter parameter = (JobParameter) jobParameter;
```

Comment:
nit: finals for the values (please check everywhere)

Reply:
will do